FoundationDB/fdb-kubernetes-operator
# Refactor reconciliation results to use a custom structure

# Description

*Please include a summary of the change and which issue is addressed. If this change resolves an issue, please include the issue number in the description.*

This introduces a new Requeue type to allow subreconcilers to return requeue results with custom messages and requeue times. It also introduces new events that get stored on clusters when reconciliation ends early.

Resolves #370 
Resolves #385

## Type of change

*Please select one of the options below.*

- New feature (non-breaking change which adds functionality)

# Discussion

*Are there any design details that you would like to discuss further?*

I chose to use a custom structure instead of `ctrl.Result`. This allows us to add custom messages that are separate from errors, and to propagate errors through the same result type. I think this makes the code for returning these results more readable.

# Testing

*Please describe the tests that you ran to verify your changes. Unit tests?
Manual testing?*

I updated the unit tests to capture scenarios where the requeue result is changing here, and ran operations in my local environment to confirm that I could see the new events.

*Do we need to perform additional testing once this is merged, or perform in a larger testing environment?*

No.

# Documentation

*Did you update relevant documentation within this repository?*

I've updated the debugging guide to reference the new event and log messages.

# Follow-up

*Are there any follow-up issues that we should pursue in the future?*

No.
